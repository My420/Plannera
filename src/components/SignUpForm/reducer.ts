import { IAuthField, ISignUpState, IAuthAction } from '../../types/signUpForm';
import {
  EMAIL, PASSWORD, CONFIRM, FIRST_NAME, LAST_NAME,
} from '../../utils/constant';
import validateEmail from '../../utils/validateEmail';
import validatePassword from '../../utils/validatePassword';
import validateConfirmField from '../../utils/validateConfirmField';
import validateNameField from '../../utils/validateNameField';

export const field: IAuthField = { value: '', isValid: false, error: '' };
export const initialState: ISignUpState = {
  [EMAIL]: { ...field },
  [PASSWORD]: { ...field },
  [CONFIRM]: { ...field },
  [FIRST_NAME]: { ...field },
  [LAST_NAME]: { ...field },
};

const reducer = (state: ISignUpState, action: IAuthAction) => {
  const {
    type,
    payload: { value },
  } = action;

  const newState = (isValid: boolean, error: string): ISignUpState => ({
    ...state,
    [type]: { value, isValid, error },
  });

  switch (type) {
    case EMAIL: {
      const { isValid, error } = validateEmail(value);
      return newState(isValid, error);
    }
    case PASSWORD: {
      const passwordStatus = validatePassword(value);
      const confirmStatus = validateConfirmField(state[CONFIRM].value, value);
      if (!passwordStatus.isValid || (passwordStatus.isValid && confirmStatus.isValid)) {
        return newState(passwordStatus.isValid, passwordStatus.error);
      }
      return {
        ...newState(passwordStatus.isValid, passwordStatus.error),
        [CONFIRM]: { ...state[CONFIRM], ...confirmStatus },
      };
    }
    case CONFIRM: {
      const { isValid, error } = validateConfirmField(value, state[PASSWORD].value);
      return newState(isValid, error);
    }
    case FIRST_NAME: {
      const { isValid, error } = validateNameField(value);
      return newState(isValid, error);
    }
    case LAST_NAME: {
      const { isValid, error } = validateNameField(value);
      return newState(isValid, error);
    }
    default:
      return state;
  }
};

export default reducer;
