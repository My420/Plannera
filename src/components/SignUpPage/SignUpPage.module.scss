@import '../../sass/variables.scss';

.page {
  width: 100vw;
  height: 100vh;
}
.title {
  text-align: center;
}
.wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;

  width: 100%;
  height: 100%;
}

.logo {
  width: 50px;
  height: 50px;
}

.appName {
  margin-bottom: 30px;
}

@media (max-width: $mobile-large) {
  .appName {
    margin-bottom: 10px;
  }
}
