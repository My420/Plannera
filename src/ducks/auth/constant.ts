import { fb as firebase } from '../../firebase/config';

export const moduleName = 'auth';
export const SIGN_UP_REQUEST = 'Plannera/auth/SIGN_UP_REQUEST';
export const SIGN_UP_SUCCESS = 'Plannera/auth/SIGN_UP_SUCCESS';
export const SIGN_UP_ERROR = 'Plannera/auth/SIGN_UP_ERROR';
export const SIGN_IN_REQUEST = 'Plannera/auth/SIGN_IN_REQUEST';
export const SIGN_IN_SUCCESS = 'Plannera/auth/SIGN_IN_SUCCESS';
export const SIGN_IN_ERROR = 'Plannera/auth/SIGN_IN_ERROR';
export const SIGN_OUT_REQUEST = 'Plannera/auth/SIGN_OUT_REQUEST';
export const SIGN_OUT_SUCCESS = 'Plannera/auth/SIGN_OUT_SUCCESS';
export const SIGN_OUT_ERROR = 'Plannera/auth/SIGN_OUT_ERROR';
export const AUTH_CLEAR_ERROR = 'Plannera/auth/AUTH_CLEAR_ERROR';

export const auth = firebase.auth();
