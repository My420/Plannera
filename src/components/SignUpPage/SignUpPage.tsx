import React from 'react';
import SignUpForm from '../SignUpForm';
import styles from './SignUpPage.module.scss';

const SignUpPage: React.FC = () => (
  <section className={styles.page}>
    <div className={styles.wrapper}>
      <h1 className={styles.title}>LOGO</h1>
      <SignUpForm />
    </div>
  </section>
);

export default SignUpPage;
