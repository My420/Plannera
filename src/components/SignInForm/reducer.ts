import { ISignInState, IAuthAction, IAuthField } from '../../types/signUpForm';
import { EMAIL, PASSWORD } from '../../utils/constant';
import validateEmail from '../../utils/validateEmail';
import validatePassword from '../../utils/validatePassword';

export const field: IAuthField = { value: '', isValid: false, error: '' };
export const initialState: ISignInState = {
  [EMAIL]: { ...field },
  [PASSWORD]: { ...field },
};

const reducer = (state: ISignInState, action: IAuthAction) => {
  const {
    type,
    payload: { value },
  } = action;

  const newState = (isValid: boolean, error: string): ISignInState => ({
    ...state,
    [type]: { value, isValid, error },
  });

  switch (type) {
    case EMAIL: {
      const { isValid, error } = validateEmail(value);
      return newState(isValid, error);
    }
    case PASSWORD: {
      const { isValid, error } = validatePassword(value);
      return newState(isValid, error);
    }
    default:
      return state;
  }
};

export default reducer;
