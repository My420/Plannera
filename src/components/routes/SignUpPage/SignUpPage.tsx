import React from 'react';
import { APP_NAME } from '../../../utils/constant';
import SignUpForm from '../../SignUpForm';
import styles from './SignUpPage.module.scss';
import Logo from '../../Logo';

const SignUpPage: React.FC = () => (
  <section className={styles.page}>
    <div className={styles.wrapper}>
      <h1 className="visually-hidden">{`${APP_NAME} sign up page`}</h1>
      <Logo wrapperClass={styles.logo} />
      <h2 className={styles.appName}>{`${APP_NAME}`}</h2>
      <SignUpForm />
    </div>
  </section>
);

export default SignUpPage;
