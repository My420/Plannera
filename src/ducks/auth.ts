import { Record } from 'immutable';
import { put, takeEvery, all } from 'redux-saga/effects';
import { ISignUpFormData, ISignInFormData } from '../types/signUpForm';
import { IUser } from '../types/user';
import { EMAIL, LAST_NAME, FIRST_NAME } from '../utils/constant';

export const moduleName = 'auth';
export const SIGN_UP_REQUEST = 'Plannera/auth/SIGN_UP_REQUEST';
export const SIGN_UP_SUCCESS = 'Plannera/auth/SIGN_UP_SUCCESS';
export const SIGN_UP_ERROR = 'Plannera/auth/SIGN_UP_ERROR';
export const SIGN_IN_REQUEST = 'Plannera/auth/SIGN_IN_REQUEST';

/**
 * types
 */

export interface ISignUpRequestAction {
  type: typeof SIGN_UP_REQUEST;
  payload: ISignUpFormData;
}
export interface ISignUpSuccessAction {
  type: typeof SIGN_UP_SUCCESS;
  payload: IUser;
}
export interface ISignUpErrorAction {
  type: typeof SIGN_UP_ERROR;
  payload: string;
}
export interface ISignInRequestAction {
  type: typeof SIGN_IN_REQUEST;
  payload: ISignInFormData;
}

export type AuthActionTypes =
  | ISignUpRequestAction
  | ISignUpSuccessAction
  | ISignUpErrorAction
  | ISignInRequestAction;

export interface IReducerInitialState extends IUser {
  error: null | string;
  loading: boolean;
}

/**
 * Reducer
 */

const initialState: IReducerInitialState = {
  userID: null,
  email: null,
  firstName: null,
  lastName: null,
  initial: null,
  error: null,
  loading: false,
};

const ReducerRecord = Record(initialState);
const reducerState = new ReducerRecord();

const reducer = (state = reducerState, action: AuthActionTypes): typeof reducerState => {
  switch (action.type) {
    case SIGN_UP_REQUEST:
      return state.set('loading', true);
    case SIGN_UP_SUCCESS: {
      const {
        userID, email, lastName, firstName, initial,
      } = action.payload;
      return state
        .set('userID', userID)
        .set('email', email)
        .set('firstName', firstName)
        .set('lastName', lastName)
        .set('initial', initial)
        .set('loading', false)
        .set('error', null);
    }
    case SIGN_UP_ERROR: {
      const error = action.payload;
      return state
        .set('userID', null)
        .set('email', null)
        .set('firstName', null)
        .set('lastName', null)
        .set('initial', null)
        .set('loading', false)
        .set('error', error);
    }
    default:
      return state;
  }
};

export default reducer;

/**
 * Action Creator
 */

export const signUp = (data: ISignUpFormData): ISignUpRequestAction => ({
  type: SIGN_UP_REQUEST,
  payload: data,
});

export const signUpSuccess = (data: IUser): ISignUpSuccessAction => ({
  type: SIGN_UP_SUCCESS,
  payload: data,
});

export const signIn = (data: ISignInFormData): ISignInRequestAction => ({
  type: SIGN_IN_REQUEST,
  payload: data,
});

/**
 *  Saga
 */

export function* registerUser(action: ISignUpRequestAction) {
  const { payload } = action;
  const { [EMAIL]: email, [FIRST_NAME]: firstName, [LAST_NAME]: lastName } = payload;

  // register new user in firebase get ID

  // create initial
  const initial = (firstName[0] + lastName[0]).toUpperCase();

  // add user profile card

  // dispatch success or error action

  const data: IUser = {
    userID: 'dsdfadsasdad2dsdsa',
    email,
    firstName,
    lastName,
    initial,
  };

  yield put(signUpSuccess(data));
}

export function* loginUser(action: ISignInRequestAction) {
  const { payload } = action;
  const { [EMAIL]: email } = payload;

  // login user in firebase get ID

  // fetch user data from firebase

  // dispatch success or error action

  const data: IUser = {
    userID: 'dsdfadsasdad2dsdsa',
    email,
    firstName: 'firstName',
    lastName: 'lastName',
    initial: 'FL',
  };

  yield put(signUpSuccess(data));
}

export function* authSaga() {
  yield all([takeEvery(SIGN_UP_REQUEST, registerUser), takeEvery(SIGN_IN_REQUEST, loginUser)]);
}
