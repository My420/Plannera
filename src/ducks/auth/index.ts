import { moduleName } from './constant';
import reducer from './reducer';
import authSaga from './saga';

export { moduleName, authSaga };
export default reducer;
