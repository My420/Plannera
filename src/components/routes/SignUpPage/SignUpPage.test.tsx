import React from 'react';
import { Router, Route } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { render, cleanup, screen } from '@testing-library/react';
import { createBrowserHistory } from 'history';
import { moduleName } from '../../../ducks/auth/constant';
import SignUpPage from './SignUpPage';
import { reducerState } from '../../../ducks/auth/reducer';
import { SIGN_UP_PAGE, APP_NAME } from '../../../utils/constant';

const mockStore = configureStore();
const history = createBrowserHistory();
history.push('/signup');

const renderWithAuthRedux = (component: React.ReactNode, initialState = reducerState) => {
  const store = mockStore({ [moduleName]: initialState });
  return {
    ...render(
      <Provider store={store}>
        <Router history={history}>
          <Route exact path={SIGN_UP_PAGE} component={component as React.FC} />
        </Router>
      </Provider>,
    ),
    store,
  };
};

describe('test SignUpPage component', () => {
  afterEach(cleanup);

  test('should match snapshot', () => {
    const { asFragment } = renderWithAuthRedux(SignUpPage);
    expect(asFragment()).toMatchSnapshot();
  });

  test('page should have hidden title for accessibility', () => {
    renderWithAuthRedux(SignUpPage);
    const element = screen.getByText(`${APP_NAME} sign up page`);
    expect(element).toBeTruthy();
  });

  test('page should have an application logo', () => {
    renderWithAuthRedux(SignUpPage);
    const element = screen.getByTitle(`${APP_NAME} logo`);
    expect(element).toBeTruthy();
  });

  test('page should show the name of the application', () => {
    renderWithAuthRedux(SignUpPage);
    const element = screen.getByText(`${APP_NAME}`);
    expect(element).toBeTruthy();
  });

  test('page should show sign in form', () => {
    renderWithAuthRedux(SignUpPage);
    const title = screen.getByText('Create account');
    const button = screen.getByText('Continue');
    expect(title).toBeTruthy();
    expect(button).toBeTruthy();
  });
});
