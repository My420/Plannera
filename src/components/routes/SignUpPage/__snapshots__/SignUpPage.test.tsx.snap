// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SignUpPage component should match snapshot 1`] = `
<DocumentFragment>
  <section
    class="page"
  >
    <div
      class="wrapper"
    >
      <h1
        class="visually-hidden"
      >
        Plannera sign up page
      </h1>
      <div
        class="logo"
      >
        <svg
          height="100%"
          viewBox="0 0 321.75 286.95"
          width="100%"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Plannera logo
          </title>
          <g
            data-testid="img"
            fill="#ececec"
            fill-rule="evenodd"
            stroke-dashoffset="2"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="4.1"
            stroke-width="10"
            transform="matrix(.95021 -.31162 .31162 .95021 -106.29 132.88)"
          >
            <path
              d="m290.76-39.011c-116.29 5.9117-207.84 67.222-207.84 141.94 0 78.622 101.38 142.44 226.28 142.44 21.72 0 42.737-1.9295 62.625-5.5312v-39.312c-17.014 3.0554-34.967 4.6875-53.531 4.6875-107.62 0-194.97-54.977-194.97-122.72 0-61.859 72.843-113.07 167.44-121.5z"
            />
            <path
              d="m304.42-37.858c-89.455 4.5475-159.88 51.71-159.88 109.18 0 60.479 77.983 109.57 174.06 109.57 16.708 0 32.875-1.4842 48.173-4.2548v-30.24c-13.088 2.3503-26.898 3.6058-41.178 3.6058-82.783 0-149.98-42.29-149.98-94.399 0-47.584 56.033-86.978 128.8-93.462z"
            />
            <path
              d="m317.45-35.731c-66.322 3.3715-118.54 38.338-118.54 80.948 0 44.839 57.817 81.234 129.05 81.234 12.387 0 24.373-1.1004 35.716-3.1545v-22.42c-9.7033 1.7425-19.942 2.6733-30.529 2.6733-61.376 0-111.19-31.354-111.19-69.988 0-35.279 41.543-64.486 95.491-69.293z"
            />
          </g>
        </svg>
      </div>
      <h2
        class="appName"
      >
        Plannera
      </h2>
      <form
        action="signup.php"
        class="form"
        name="signUpForm"
      >
        <p
          class="title"
        >
          Create account
        </p>
        <div
          class="status"
        >
          <div
            class="container"
            data-testid="container"
          />
        </div>
        <fieldset
          class="signup"
        >
          <label
            class="label input"
            for="email"
          >
            <span
              class="visually-hidden"
            >
              Email
            </span>
            <span
              class="error"
              data-testid="error"
            />
            <input
              class="input "
              id="email"
              name="EMAIL"
              placeholder="Email"
              type="email"
              value=""
            />
          </label>
          <label
            class="label input"
            for="password"
          >
            <span
              class="visually-hidden"
            >
              Password
            </span>
            <span
              class="error"
              data-testid="error"
            />
            <input
              class="input "
              id="password"
              name="PASSWORD"
              placeholder="Password"
              type="password"
              value=""
            />
          </label>
          <label
            class="label input"
            for="confirm"
          >
            <span
              class="visually-hidden"
            >
              Confirm password
            </span>
            <span
              class="error"
              data-testid="error"
            />
            <input
              class="input "
              id="confirm"
              name="CONFIRM"
              placeholder="Confirm password"
              type="password"
              value=""
            />
          </label>
          <label
            class="label input"
            for="firstName"
          >
            <span
              class="visually-hidden"
            >
              First name
            </span>
            <span
              class="error"
              data-testid="error"
            />
            <input
              class="input "
              id="firstName"
              name="FIRST_NAME"
              placeholder="First name"
              type="text"
              value=""
            />
          </label>
          <label
            class="label input"
            for="lastName"
          >
            <span
              class="visually-hidden"
            >
              Last name
            </span>
            <span
              class="error"
              data-testid="error"
            />
            <input
              class="input "
              id="lastName"
              name="LAST_NAME"
              placeholder="Last name"
              type="text"
              value=""
            />
          </label>
        </fieldset>
        <fieldset
          class="controls"
        >
          <button
            class="button "
            disabled=""
            name="signUpSubmit"
            type="submit"
          >
            Continue
          </button>
          <a
            class="link"
            href="/signin"
          >
            Sign in
          </a>
        </fieldset>
      </form>
    </div>
  </section>
</DocumentFragment>
`;
